24.md

```diff
@@ -0,0 +1,395 @@
+      "created_at":1690558348,
+      "id":"96b5def1d84c5c33d90fc3a36133a32eba0eca98dd67f630d442a2456d5b9c4e",
+      "kind":1059,
+      "pubkey":"1c1b8ab402dcdf7bdaeba8cf5056b2ec5d2958886bcaf342cd37d71cfde1bf8d",
+      "sig":"1144271fc0e3d07df0ba72b58a4544a469e83314bf29dc9275ea67d0d0a7ec4989487fdcf1ae9958710e0956d56a3833e92ce40dd22f0de628718888538d098f",
+      "tags":[
+        [ "p", "2d5ad8fb75c30aeb153e6d8b6a7275424be54d196f4837d658cdd36c81dad8d3" ]
+      ]
+    }
+    ```
+
+   5.3 Broadcast the `kind:1059` to the sender's relays
+
+## 9. Encryption Code Examples
+
+### 9.1. JavaScript
+
+```js
+// npm install @noble/curves @noble/ciphers @noble/hashes @scure/base @stablelib/xchacha20
+import {xchacha20} from '@noble/ciphers/chacha'
+import {secp256k1} from '@noble/curves/secp256k1'
+import {sha256} from '@noble/hashes/sha256'
+import {randomBytes} from '@noble/hashes/utils'
+import {base64} from '@scure/base'
+
+
+const utf8Decoder = new TextDecoder()
+const utf8Encoder = new TextEncoder()
+
+export function getConversationKey(privkeyA: string, pubkeyB: string): Uint8Array {
+  const key = secp256k1.getSharedSecret(privkeyA, '02' + pubkeyB)
+  return sha256(key.subarray(1, 33))
+}
+
+export function encrypt(
+  key: Uint8Array,
+  text: string,
+  ver = 1
+): string {
+  if (ver !== 1) throw new Error('NIP44: unknown encryption version')
+  let nonce = randomBytes(24)
+  let plaintext = utf8Encoder.encode(text)
+  let ciphertext = xchacha20(key, nonce, plaintext, plaintext)
+  return `1,${base64.encode(nonce)},${base64.encode(ciphertext)}`
+}
+
+export function decrypt(key: Uint8Array, data: string): string {
+  let dt = data.split(',')
+  if (dt.length !== 3) throw new Error('NIP44: unknown encryption version');
+  let v = Number.parseInt(dt[0])
+  if (v !== 1) throw new Error('NIP44: unknown encryption version')
+
+  let nonce = base64.decode(dt[1])
+  let ciphertext = base64.decode(dt[2])
+  let plaintext = xchacha20(key, nonce, ciphertext)
+  let text = utf8Decoder.decode(plaintext)
+  return text
+}
+```
+
+### 9.2. Kotlin
+  
+```kotlin
+
+import com.goterl.lazysodium.LazySodiumAndroid
+import com.goterl.lazysodium.SodiumAndroid
+import com.goterl.lazysodium.interfaces.AEAD
+import com.goterl.lazysodium.utils.Base64MessageEncoder
+import com.goterl.lazysodium.utils.Key
+import java.util.Base64
+
+data class EncryptedInfo(val ciphertext: String, val nonce: String, val v: Int)
+
+enum class Nip44Version(val versionCode: Int) {
+  Reserved(0),
+  XChaCha20(1)
+}
+
+fun encryptXChaCha(msg: String, sharedSecret: ByteArray): EncryptedInfo {
+  val lazySodium = LazySodiumAndroid(SodiumAndroid(), Base64MessageEncoder())
+
+  val key = Key.fromBytes(sharedSecret)
+
+  val nonce: ByteArray = lazySodium.nonce(24)
+  val cipherBase64: String = lazySodium.encrypt(msg, null, nonce, key, AEAD.Method.XCHACHA20_POLY1305_IETF)
```

Comment:
Is that correct? `XCHACHA20_POLY1305`?

Reply:
Frankly, I am still trying to figure it out... I don't know if the JS version can be decrypted by the Kotlin version yet.

Reply:
Use this: https://github.com/terl/lazysodium-java/blob/64dc78554f45eb9b81e74354d61eba4e3da2452b/src/main/java/com/goterl/lazysodium/SodiumJava.java#L139

Reply:
I don't have that method on Android. It's only available on desktop. :( 

Is it the only way?

Reply:
i've thought the lazysodium-android library exposes all lazysodium methods? Are you 100% sure you can't access it?

Yes -- we need xchacha20, not xchacha20poly1305, they return different results.

Reply:
No, they are different implementations. The Android version doesn't have xchacha20XOR. The author said they were having lots of issues with that call in native android. :(

Reply:
Please @vitorpamplona have you tried out this library: https://github.com/ionspin/kotlin-multiplatform-libsodium ? (The only issue is the experimental status of the lib :sweat_smile: )